File: book_this-is-coding-test-with-python/part2/chap9-shortest_path/dijkstra-V2/solution.py
import sys

input = sys.stdin.readline
INF = int(1e9)

n, m = map(int, input().split())
start = int(input())

graph = [[] for i in range(n + 1)]
visited = [False] * (n + 1)
distance = [INF] * (n + 1)

for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))


def get_smallest_node():
    min_value = INF
    index = 0
    for i in range(1, n + 1):
        if distance[i] < min_value and not visited[i]:
            min_value = distance[i]
            index = i
    return index


def dijkstra(start):
    # mine
    # distance[start] = 0
    # target_node = start

    # while target_node != 0:
    #     visited[target_node] = True
    #     adjacent_nodes = graph[target_node]
    #     for node, dist in adjacent_nodes:
    #         distance[node] = min(distance[node], distance[target_node] + dist)
    #     target_node = get_smallest_node()
    distance[start] = 0
    visited[start] = True
    for i in graph[start]:
        distance[i[0]] = i[1]

    for _ in range(n - 1):
        now = get_smallest_node()
        visited[now] = True
        for i in graph[now]:
            cost = distance[now] + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost


dijkstra(start)

for i in distance[1:]:
    if i == INF:
        print("INFINITY", end=" ")
    else:
        print(i, end=" ")

File: book_this-is-coding-test-with-python/part2/chap9-shortest_path/dijkstra-V2/solution.js
const INF = 1000000000;
const n = 6;
const m = 11;
const start = 1;
const graph = [
  [],
  [
    [2, 2],
    [3, 5],
    [4, 1],
  ],
  [
    [3, 3],
    [4, 2],
  ],
  [
    [2, 3],
    [6, 5],
  ],
  [
    [3, 3],
    [5, 1],
  ],
  [
    [3, 1],
    [6, 2],
  ],
  [],
];
const visited = new Array(n + 1).fill(false);
const distance = new Array(n + 1).fill(INF);

function getSmallestNode() {
  let minValue = INF;
  let index = 0;

  for (let i = 1; i <= n; i++) {
    if (visited[i] === false && distance[i] < minValue) {
      minValue = distance[i];
      index = i;
    }
  }

  return index;
}

function dijkstra(start) {
  distance[start] = 0;
  visited[start] = true;

  for (const adjacent of graph[start]) {
    const [node, cost] = adjacent;
    distance[node] = cost;
  }

  for (let i = 0; i < n - 1; i++) {
    const now = getSmallestNode();
    visited[now] = true;

    for (const adjacent of graph[now]) {
      const [node, cost] = adjacent;
      const totalCost = distance[now] + cost;
      if (totalCost < distance[node]) {
        distance[node] = totalCost;
      }
    }
  }
}

dijkstra(start);

for (let i = 1; i <= n; i++) {
  if (distance[i] === INF) console.log('INF ');
  else console.log(distance[i]);
}
